File: src/DataScience/Octave/flyweight.m
function flyweight()
  pool = containers.Map();

  [pool, red1] = get_style(pool, 'Inter', 12, 'red');
  [pool, red2] = get_style(pool, 'Inter', 12, 'red');
  [pool, blue] = get_style(pool, 'Inter', 12, 'blue');

  assert(strcmp(blue.color, 'blue'));
  shared = strcmp(red1.font, red2.font) && red1.size == red2.size && strcmp(red1.color, red2.color);
  fprintf('styles=%d;shared=%s;text=ABC\n', pool.Count, lower(mat2str(shared)));
end

function [pool, style] = get_style(pool, font, size, color)
  key = sprintf('%s|%d|%s', font, size, color);
  if ~isKey(pool, key)
    pool(key) = struct('font', font, 'size', size, 'color', color);
  end
  style = pool(key);
end

flyweight();

File: src/Niche/Prolog/flyweight.pl
:- initialization(main, main).

intern_style(Key, Pool, Pool, Id) :-
    memberchk(Key-Id, Pool), !.
intern_style(Key, Pool0, [Key-Id|Pool0], Id) :-
    length(Pool0, Count),
    Id is Count + 1.

main :-
    intern_style(style('Inter', 12, red), [], Pool1, Red1),
    intern_style(style('Inter', 12, red), Pool1, Pool2, Red2),
    intern_style(style('Inter', 12, blue), Pool2, Pool3, Blue),
    Blue =:= 2,
    length(Pool3, Styles),
    ( Red1 =:= Red2 -> Shared = true ; Shared = false ),
    format('styles=~d;shared=~w;text=ABC~n', [Styles, Shared]).
